### ISSUE TITLE
Why do we use the current `_version.py` approach for our `setup.py` files?

### ISSUE THREAD

**Issue by christophertubbs** [NOAA-OWP/DMOD#423]
Our `setup.py` files generally always have a line like:

```python
exec(open(ROOT / 'dmod/evaluationservice/_version.py').read())
```

Instead of having `_version.py` declared next to `setup.py` and importing the value or just declaring the value itself in the `setup.py`, we're getting that value under the nested directory, which causes your IDE to become grumpy and can cause testing issues through github actions if someone forgets to update it.

Why not evaluate the version programmatically by combining the tag of the current repo with some derivative of git's `hash-object` function over `ROOT` (maybe the first 4 characters)? This would keep everything up to date with new changes, create a more consistent versioning scheme, and allow packages to be rebuilt and tested whenever changes are made within, even if someone forgets to update their `_version.py`? Given that we don't have any tags, if I went with 7 characters for my branch version, I'd get something like  `v0.0.0-4196e18` for the `evaluation_service`'s version that I'm currently working with (if `v0.0.0` is the default tag).

**Comment by robertbartel** [comment]
> Why not evaluate the version programmatically by combining the tag of the current repo with some derivative of git's hash-object function over ROOT (maybe the first 4 characters)? 

@christophertubbs, I think the answer to this is because it would require us to use strict version clauses for package dependencies. We couldn't get away with depending on `>=0.8.0` for example. Unless there is something about the hashing algorithm I'm forgetting about.

**Comment by robertbartel** [comment]
FWIW, my above comment is not to say we shouldn't explore alternatives to our current versioning scheme. I just think that suggestion would cause more disruption than it removes.

**Comment by christophertubbs** [comment]
And... that's why I marked it as a 'question'. Not cut and dry, definitely not something that requires immediate action. Just an oddity that I believe needs to be written down before the confusion is forgotten.

**Comment by aaraney** [comment]
> Instead of having `_version.py` declared next to `setup.py` and importing the value or just declaring the value itself in the `setup.py`, we're getting that value under the nested directory, which causes your IDE to become grumpy and can cause testing issues through github actions if someone forgets to update it.

@christophertubbs, first off, sorry to hear you are having issues. That sounds annoying. To continue the conversation, can you elaborate a bit on the grips your IDE is having with the current approach?

To add a little context, the reason that `_version.py` files are defined within a given package (e.g. `dmod.evaluationservice._version.py`) and then read by `setup.py` is so version information is accessible during _run time_ and _build time_. If we did not specify the version inside the package, then the only way to determine which version of the package you have would be by using `pip` or some other package management tool.

> Do you think there's a way to build a git action to say "Oh, you changed stuff under this directory - did you change the `_version.py`? Nope? Fail.

I don't see why not. I am curious if we could write a custom plugin for [`pre-commit`](https://pre-commit.com/) that checks this and then just run `pre-commit` in our CI. Thoughts?

**Comment by christophertubbs** [comment]
You're the genius, you tell me :)

I generally have trouble setting up git actions, so I don't think I could whip up a solution.

Out of curiosity, how does the library version fit into runtime workflows? i.e. why do we need `__version__` to be accessible in-app and why should I be sticking it into my own applications? Sounds like I've got a blind spot.

**Comment by aaraney** [comment]
> I generally have trouble setting up git actions, so I don't think I could whip up a solution.

Yeah, Ill have to think about what that might look like. I think this ultimately boils down to our release and versioning strategy. My gut says that given the size of our team each PR should include a version bump of some kind (e.g. patch). If this is causing enough friction, we should talk about this as a group and layout some expectations. Thoughts?

> Out of curiosity, how does the library version fit into runtime workflows? i.e. why do we need `__version__` to be accessible in-app and why should I be sticking it into my own applications? Sounds like I've got a blind spot.

I don't know that it _fits into_ runtime workflows. The main purpose is to have the version of the code in the code. Since python is interpreted this means sticking a string literal into you package, where in other language this would get included in the binary as metadata. You can imagine this being helpful when debugging or reporting errors. [`pandas.show_versions()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.show_versions.html) is a great example of this. Additionally with `python` virtual environments and possible collisions with the `PATH` environment variable when running `pip`, it is really nice to get the version from the code in a repl session for example.
